### PR TITLE
Unifying and elevating EOS functionality in Enzo-E

### PR DESCRIPTION
**Pull request by mabruzzo**
This PR finally introduces some changes that I've been working towards for about 2 years now.

This PR does 2 things:
1. It makes the EOS a global property of the simulation that is configured through the ``"fluid_props"`` physics object (previously we implicitly assumed an ideal-gas eos everywhere other than the ppml solver). 
    - The main point is to try to have a single source of truth where various EOS-related functionality is implemented in the C++ codebase.
    - The secondary goal is to make it easy for various Methods to either 
        - easily state their assumptions what kind of EOS is used (and cause the code to exit with an error when that is violated)
        - customize their behavior based on the specified type of EOS

2. Previously the VL+CT solver had implemented 2 disconnected versions of the EOS. This PR essentially removed the version with the class-based inheritance. Now we use the version that was employed within the Riemann Solvers everywhere.

~~This PR is 95% done. I really need just need to finish writing up the documentation for this. In particular, I think the way that I've opted to implement polymorphism is probably un-intuitive for most people, and the documentation will justify why we adopt this strategy (and explain why it is the superior choice to class-based inheritance).~~

~~Once I've finished with that, I update this PR to indicate it's no longer a WIP~~

EDIT: This is now ready for review. The website has been updated with a bunch of new documentation. We need to update the gold standard to reflect the recent changes (but that needs to be handled after review)

### CODE REVIEW

File: src/Enzo/enzo_EnzoMethodPpml.cpp
Comment (bwoshea) [inline src/Enzo/enzo_EnzoMethodPpml.cpp:32]: It should probably say PPML instead of PPM in this entire code block, right?

File: src/Enzo/enzo_EnzoPhysicsFluidProps.cpp
Comment (bwoshea) [inline src/Enzo/enzo_EnzoPhysicsFluidProps.cpp:29]:
```suggestion
  //          "can't return gamma for EOS of type \"%s\". Either that type "
```

File: src/Enzo/enzo_EnzoPhysicsFluidProps.cpp
Comment (bwoshea) [inline src/Enzo/enzo_EnzoPhysicsFluidProps.cpp:41]:
```suggestion
         "can't return gamma for EOS of type \"%s\". Either that type "
```

File: src/Enzo/enzo_EnzoPhysicsFluidProps.hpp
Comment (bwoshea) [inline src/Enzo/enzo_EnzoPhysicsFluidProps.hpp:70]:
```suggestion
  /// When the stored EOS variant doesn't contain a value of gamma, the program
```

**Review by bwoshea — CHANGES_REQUESTED** [review]
Overall the code looks great to me and is a good thing to have in Enzo-E.  I have a bunch of very minor suggestions, mostly relating to the documentation and so forth.  I'm happy to approve once you make the requested updates!

**Review by dcollins4096 — APPROVED** [review]

File: src/Enzo/enzo_EnzoMethodPpml.cpp
Comment (mabruzzo) [reply src/Enzo/enzo_EnzoMethodPpml.cpp:32]: You're absolutely right. I'll fix that

File: doc/source/devel/eos-fluidprops.rst
Comment (mabruzzo) [reply doc/source/devel/eos-fluidprops.rst:370]: The discussion of the pressure field was very confusing (it was a half-baked thought). I have replaced the second paragraph with the following blurb. Let me know if this addressed your concerns.

```rst
The following code snippet shows a somewhat simplified example of how
you might perform this calculation.  This snippet will abort with an
error if each Processing Element's global
:cpp:class:`!EnzoPhysicsFluidProps` instance was configured to hold
anything other than an ideal EOS.
```

File: doc/source/devel/eos-fluidprops.rst
Comment (mabruzzo) [reply doc/source/devel/eos-fluidprops.rst:385]: The short answer is: NOT YET. It's absolutely possible and it was originally a major motivation for introducing the ``visit`` function. However, in practice it's not worth doing right now when we only have two equations of state and while we are using C++11.

Here are 2 examples (while they are possible in C++11, I've opted not to show that right now because that is more cumbersome).

If every EOS class has a method with the signature ``enzo_float EOSClass::sound_speed(enzo_float density, enzo_float pressure)``, then we can write easily write the following snippet in C++14 that will do this:

```c++

   double timestep(CelloArray<const enzo_float, 3> density,
                   CelloArray<const enzo_float, 3> velocity_x,
                   CelloArray<const enzo_float, 3> velocity_y,
                   CelloArray<const enzo_float, 3> velocity_z,
                   CelloArray<const enzo_float, 3> pressure,
                   double dx, double dy, double dz,
                   double courant_factor)
   {
     auto lambda_func = [=](auto eos) -> double
     {
       const int mx = density.shape(2);
       const int my = density.shape(1);
       const int mz = density.shape(0);

       double dt = std::numeric_limits<double>::max();
       for (int iz = 0; iz < mz; iz++) {
         for (int iy = 0; iy < my; iy++) {
           for (int ix = 0; ix < mx; ix++) {

             double cs = (double) eos.sound_speed(density(iz,iy,ix),
                                                  pressure(iz,iy,ix));
             double abs_vx = std::fabs((double) velocity_x(iz,iy,ix));
             double abs_vy = std::fabs((double) velocity_y(iz,iy,ix));
             double abs_vz = std::fabs((double) velocity_z(iz,iy,ix));
             double tmp = std::min(std::min(dx/(abs_vx + cs),
                                            dy/(abs_vy + cs)),
                                   dz/(abs_vz + cs));
             dt = std::min(dt, tmp);
           }
         }
       }
       return dt;
     };
     double dt = enzo::fluid_props()->eos_variant().visit(lambda_func);
     return courant_factor * dt;
    }
```

In practice, it doesn't make sense for every EOS to have a method that computes the sound speed from both density and pressure. In the future, one could imagine modifying ``EnzoEOSIsothermal`` so that it has the method ``enzo_float EnzoEOSIsothermal::sound_speed()``. In C++17 we could write a new function that makes use of that new functionallity:

```c++

   double timestep(CelloArray<const enzo_float, 3> density,
                   CelloArray<const enzo_float, 3> velocity_x,
                   CelloArray<const enzo_float, 3> velocity_y,
                   CelloArray<const enzo_float, 3> velocity_z,
                   CelloArray<const enzo_float, 3> pressure,
                   double dx, double dy, double dz,
                   double courant_factor)
   {
     auto lambda_func = [=](auto eos) -> double
     {
       using T = std::decay_t<decltype(arg)>;
       const int mx = density.shape(2);
       const int my = density.shape(1);
       const int mz = density.shape(0);

       double dt = std::numeric_limits<double>::max();
       for (int iz = 0; iz < mz; iz++) {
         for (int iy = 0; iy < my; iy++) {
           for (int ix = 0; ix < mx; ix++) {

             double cs;

             if constexpr (std::is_same_v<T, EnzoEOSIdeal>) {
               cs = (double) eos.sound_speed(density(iz,iy,ix),
                                             pressure(iz,iy,ix));
             } else if constexpr (std::is_same_v<T, EnzoEOSIsothermal>) {
               cs = (double) eos.sound_speed();
             } else {
               static_assert(false, "can't handle specified EOS");
             }

             double 
             double abs_vx = std::fabs((double) velocity_x(iz,iy,ix));
             double abs_vy = std::fabs((double) velocity_y(iz,iy,ix));
             double abs_vz = std::fabs((double) velocity_z(iz,iy,ix));
             double tmp = std::min(std::min(dx/(abs_vx + cs),
                                            dy/(abs_vy + cs)),
                                   dz/(abs_vz + cs));
             dt = std::min(dt, tmp);
           }
         }
       }
       return dt;
     };
     double dt = enzo::fluid_props()->eos_variant().visit(lambda_func);
     return courant_factor * dt;
    }
```

Note that the if-statements in the for-loop are called "constexpr-if-statements". As I understand it, they will not produce branching within the for-loop; instead separate versions of the lambda expression are automatically generated for each possible EOS type (so effectively there is a branch outside of the if-statement).

File: doc/source/devel/eos-fluidprops.rst
Comment (mabruzzo) [reply doc/source/devel/eos-fluidprops.rst:228]: Gotcha. I've fixed that

**Comment by mabruzzo** [comment]
@bwoshea This is ready for you to take another look at.

The failing single-precision answer test is expected. Once you approve the PR, I will take the necessary steps to update the answer tests reference results

**Comment by mabruzzo** [comment]
Thanks! Merging now!

**Pull request merged**
